samcli/commands/local/lib/intrinsic_resolver/intrinsic_property_resolver.py

```diff
@@ -0,0 +1,759 @@
+    FN_TRANSFORM = "Fn::Transform"
+    FN_GET_AZS = "Fn::GetAZs"
+    REF = "Ref"
+    FN_GET_ATT = "Fn::GetAtt"
+    FN_IMPORT_VALUE = "Fn::ImportValue"
+
+    SUPPORTED_INTRINSIC_FUNCTIONS = [
+        FN_JOIN,
+        FN_SPLIT,
+        FN_SUB,
+        FN_SELECT,
+        FN_BASE64,
+        FN_FIND_IN_MAP,
+        FN_TRANSFORM,
+        FN_GET_AZS,
+        REF,
+        FN_GET_ATT,
+        FN_IMPORT_VALUE
+    ]
+
+    FN_AND = "Fn::And"
+    FN_OR = "Fn::Or"
+    FN_IF = "Fn::If"
+    FN_EQUALS = "Fn::Equals"
+    FN_NOT = "Fn::Not"
+
+    CONDITIONAL_FUNCTIONS = [
+        FN_AND,
+        FN_OR,
+        FN_IF,
+        FN_EQUALS,
+        FN_NOT
+    ]
+
+    def __init__(self, symbol_resolver,
+                 template=None,
+                 resources=None,
+                 mappings=None,
+                 parameters=None,
+                 conditions=None):
+        """
+        Initializes the Intrinsic Property class. Customers can pass in their own ref_resolver or attribute_resolver
+        to customize the behavior of ref resolution. This can also be done by extending the definition.
+
+        When created the class can either process a template or the individual attributes. The more information given
+        the easier resolution will be.
+
+        In the future, for items like Fn::ImportValue multiple templates can be provided
+        into the function.
+        Parameters
+        -----------
+        symbol_resolver: IntrinsicsSymbolTable
+            This is a required property that is the crux of the intrinsics resolver. It acts as a mapping between
+            resource_type and object to the Ref, Arn, etc.
+        template: dict
+            A dictionary containing all the resources and the attributes of the CloudFormation template.
+            Customers can pass in the template or the individual properties like resources, mapping, parameters,
+            conditions.
+        resources: dict
+            Dictionary containing the CloudFormation resources
+        mappings: dict
+            Dictionary containing the CloudFormation mappings. This is used in the Fn::FindInMap section
+        parameters: dict
+            Dictionary containing the CloudFormation parameters. This is used in the Ref section
+        conditions: dict
+            Dictionary containing the CloudFormation conditions. This is used in the Boolean Intrinsics section
+        """
+        self.template = template or {}
+        self.resources = resources or self.template.get("Resources", {})
+        self.mapping = mappings or self.template.get("Mappings", {})
+        self.parameters = parameters or self.template.get("Parameters", {})
+        self.conditions = conditions or self.template.get("Conditions", {})
+
+        self.symbol_resolver = symbol_resolver
+        self.symbol_resolver.resources = self.resources
+        self.symbol_resolver.parameters = self.parameters
+
+        self.intrinsic_key_function_map = {
```

Comment:
This seems like another class that could be created? Any reason you elected to do this mapping vs. making it into an explicit class?

Reply:
It seemed nicer in terms of it's functional formatting and structure.